### PR TITLE
Implement hpct-managers library and add sync statuses

### PR DESCRIPTION
**Pull request by NucciTheBoss**
This pull request makes the following changes:

* Use new `hpct-managers` library.
* Have local manager classes inherit from `UbuntuManager`.
* Drop `apt` and `systemd` charm libraries since functionality is now included in `UbuntuManager` and `SystemdManager`.
* Add sync statuses for `slurm-client-ready`, `auth-munge`, `slurm-compute`, and `slurm-controller` relations.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -54,6 +57,11 @@ def __init__(self, *args) -> None:
             self.on.slurm_controller_relation_changed, self._slurm_controller_relation_changed
         )
 
+        self.service_init_sync("auth-munge", False)
+        self.service_init_sync("slurm-compute", False)
+        self.service_init_sync("slurm-controller", False)
```
Comment (j4m-can) [inline src/charm.py]: Do you want to track these?

File: src/manager/munge.py [hunk 2]
```diff
@@ -54,119 +45,28 @@ def generate_new_key(self) -> None:
...
-            raise MungeManagerError("Munge is not installed.")
-
-    def __is_installed(self) -> bool:
-        """Internal function to check in munge Debian package is installed on the unit.
-
-        Returns:
-            bool: True if Debian package is present; False if Debian package is not present.
-        """
-        return True if apt.DebianPackage.from_installed_package("munge").present else False
+            raise ManagerException("Munge is not installed.")
```
Comment (j4m-can) [inline src/manager/munge.py:72]: `UbuntuManager.restart() would handle this.

File: src/manager/slurm_client.py [hunk 3]
```diff
@@ -63,109 +53,16 @@ def get_hash(self, path: Union[str, None] = None) -> Union[str, None]:
...
-            raise SlurmClientManagerError("slurmd is not installed.")
-
-    def __is_installed(self) -> bool:
-        """Internal function to check if slurmd Debian package is installed on the unit.
-
-        Returns:
-            bool: True if Debian package is present; False if Debian package is not present.
-        """
-        return True if apt.DebianPackage.from_installed_package("slurmd").present else False
+            raise ManagerException("slurmd is not installed.")
```
Comment (j4m-can) [inline src/manager/slurm_client.py:68]: `UbuntuManager.restart() would handle this.

**Review by j4m-can — CHANGES_REQUESTED** [review]
Please review.

I want to see your answer about some of the syncs before commenting further.

File: src/charm.py (hunk 1 above)
Comment (NucciTheBoss) [reply src/charm.py]: To be honest, I am not sure. These relations are needed to make the server and client work correctly even after the subordinate relation is ready. From what I gather from syncs, they are supposed to tell you when relations are up to date, but that might not be the case.

Perhaps `stale` is the better option to use?

Comment (j4m-can) [reply src/charm.py]: For the `slurm-compute`, I don't see having a sync being useful.

On the client side, these would indicate if the local controller config and munge key were loaded. Although the relation is involved in updating these syncs, they are not about the relation but the data being pulled from the relation and stored. So, in the relation-changed handlers, the sync would be set to `False`, and no actual update of the config just then.

At some time later, when the admin allows for the daemon to be restarted, the config and key can be pulled and stored locally and the syncs updated to `True`.

The stale setting would be important here, but let me get back to you on it. I need to check something first.

**Review by j4m-can — CHANGES_REQUESTED** [review]
Almost there.

**Review by jnsgruk — CHANGES_REQUESTED** [review]
I'm not sure this change to do package management outside of the charm lib is appropriate. For a start, the `apt` Python library has a binary dependency, which is likely to cause build issues down the line. Having looked at the `hpct-managers` code, its logic for determining installed debs is also a little flawed.

I'd prefer we stayed consistent with the rest of the ecosystem and used the charm libs for apt/systemd/snap functionality. If there is something missing from those libraries, lets discuss.

N.B. Having read further, the charm code would be much more explicit about what's actually happening with systemd if you indeed stuck with the charm lib -- there is a fair amount of looping logic etc. in the systemd manager.

**Comment by j4m-can** [comment]
I will take the merge request for now and will revisit once Jon and I speak.

BTW, I have copied to an email Jon's comment above so as to not lose track of it because it must be addressed.

**Pull request merged**